Added Kronecker product of tensors (torch.kron)

IvanYashchuk opened the pull request:
This PR adds a function for calculating the Kronecker product of tensors.
The implementation is based on `at::tensordot` with permutations and reshape.
Tests pass.

TODO:

- [x] Add more test cases
- [x] Write documentation
- [x] Add entry `common_methods_invokations.py`

Ref. #42666

IvanYashchuk committed:
Added torch.kron

Tests pass. The implementation is based on tensordot.

IvanYashchuk committed:
Updated tests

IvanYashchuk committed:
Now kron permutation and reshape is correct

IvanYashchuk committed:
Rewrote using ternary

IvanYashchuk committed:
flake8

IvanYashchuk committed:
Added documentation

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-kron

IvanYashchuk committed:
Added fp32, complex dtypes

IvanYashchuk committed:
Added overrides entry for torch.kron

vishwakftw reviewed (commented):
Overall, looks good to me. Can we get some benchmarks too - comparison to `np.linalg.kron`?

IvanYashchuk commented:
[Here is the code for the benchmark.](https://gist.github.com/IvanYashchuk/23118e5ba9d266899fe9ee4f0fbac1b6)
The first column shows the input shapes.
```
Using pytorch 1.7.0a0+87f98a5
Using cupy 8.0.0
Using numpy 1.19.1
[------------------------------------------------ kron torch.float64 -----------------------------------------------]
                                          |  torch.kron CUDA  |  cupy.kron CUDA  |  torch.kron CPU  |  numpy.kron CPU
8 threads: ----------------------------------------------------------------------------------------------------------
      ((32,), (2, 16, 32, 32))            |        286.9      |        237.7     |       5744.7     |       5123.1
      ((32,), (2, 16, 32))                |        261.3      |        180.8     |        857.5     |        103.1
      ((32,), (32, 32))                   |        260.1      |        153.0     |        859.3     |         92.0
      ((32,), (32,))                      |        176.7      |        115.9     |        128.7     |         45.7
      ((32, 32), (2, 16, 32, 32))         |       1526.0      |       4595.9     |     262815.6     |     302219.1
      ((32, 32), (2, 16, 32))             |        279.3      |        274.7     |       5813.7     |       4914.2
      ((32, 32), (32, 32))                |        263.0      |        242.8     |       5952.8     |       4214.5
      ((32, 32), (32,))                   |        176.8      |        237.0     |        141.7     |        125.1
      ((2, 16, 32), (2, 16, 32, 32))      |       1524.0      |       4535.8     |     287108.8     |     370151.5
      ((2, 16, 32), (2, 16, 32))          |        274.9      |        277.6     |       6203.5     |       5548.5
      ((2, 16, 32), (32, 32))             |        266.5      |        264.0     |       5987.0     |       5826.5
      ((2, 16, 32), (32,))                |        192.9      |        287.9     |        156.6     |        147.4
      ((2, 16, 32, 32), (2, 16, 32, 32))  |      48484.6      |     143183.6     |    6715202.0     |    9877618.7
      ((2, 16, 32, 32), (2, 16, 32))      |       1434.0      |       4537.3     |     265152.6     |     368572.5
      ((2, 16, 32, 32), (32, 32))         |       1434.5      |       4534.6     |     257442.4     |     299326.7
      ((2, 16, 32, 32), (32,))            |        192.1      |        407.5     |        814.7     |       9662.0

[------------------------------------------------ kron torch.float32 -----------------------------------------------]
                                          |  torch.kron CUDA  |  cupy.kron CUDA  |  torch.kron CPU  |  numpy.kron CPU
8 threads: ----------------------------------------------------------------------------------------------------------
      ((32,), (2, 16, 32, 32))            |        283.1      |       233.4      |       3659.7     |       2113.9
      ((32,), (2, 16, 32))                |        261.1      |       187.5      |        610.7     |         80.2
      ((32,), (32, 32))                   |        259.9      |       159.9      |        615.1     |         74.6
      ((32,), (32,))                      |        179.2      |       116.9      |        125.8     |         45.0
      ((32, 32), (2, 16, 32, 32))         |        727.8      |      2346.8      |     151031.6     |     169541.6
      ((32, 32), (2, 16, 32))             |        282.8      |       300.1      |       3959.2     |       2814.5
      ((32, 32), (32, 32))                |        273.9      |       240.1      |       3709.1     |       1576.1
      ((32, 32), (32,))                   |        192.5      |       275.5      |        147.8     |        123.0
      ((2, 16, 32), (2, 16, 32, 32))      |        733.9      |      2297.3      |     145282.7     |     179567.1
      ((2, 16, 32), (2, 16, 32))          |        274.9      |       285.3      |       4084.3     |       2354.8
      ((2, 16, 32), (32, 32))             |        299.1      |       295.3      |       3974.9     |       2112.1
      ((2, 16, 32), (32,))                |        192.4      |       281.5      |        145.7     |        114.7
      ((2, 16, 32, 32), (2, 16, 32, 32))  |      23096.7      |     71982.7      |    4302047.3     |    5118863.5
      ((2, 16, 32, 32), (2, 16, 32))      |        711.2      |      2353.8      |     142618.3     |     181621.0
      ((2, 16, 32, 32), (32, 32))         |        713.4      |      2347.4      |     140499.8     |     161035.3
      ((2, 16, 32, 32), (32,))            |        201.4      |       446.2      |        327.4     |       3295.2

Times are in microseconds (us).
```

vishwakftw commented:
Any idea why some cases are not as fast?

IvanYashchuk commented:
Well, the implementation is different. NumPy's implementation is based on `outer` ([link to code](https://github.com/numpy/numpy/blob/v1.19.0/numpy/lib/shape_base.py#L1154)), while here `tensordot` is used. `torch.outer` is for vectors only and probably would require more manipulations to get it right.

IvanYashchuk commented:
Alright, I've realized that the previous timings were in debug mode 😄
I've updated the previous post. Now we see that the current implementation is faster than NumPy and CuPy.

IvanYashchuk committed:
Renamed a, b -> self, other

IvanYashchuk committed:
Added out= variant

IvanYashchuk committed:
Updated documentation: added note on real and complex support, added out kwarg

IvanYashchuk committed:
Added empty and out= test cases

IvanYashchuk committed:
Added entry to common_methods_invocations.py

IvanYashchuk committed:
Added test for error with incorrect out=

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-kron

IvanYashchuk committed:
Added tensor_docs entry

vishwakftw commented:
@IvanYashchuk looks good to me, thanks for this!

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1689,5 +1689,94 @@ Tensor chain_matmul(TensorList matrices) {
+        : other_sizes[ndim_diff + i] * self_sizes[i];
+  }
+
+  TORCH_CHECK(result.sizes().equals(expected_result_shape),
```
Why not just call the function and then check that result has the desired shape?

IvanYashchuk replied:
I thought that the actual computations might be expensive while calculating the expected shape (even twice) should be cheap, so it might be a good idea to raise exception earlier if that's possible.

In the future, this should be changed that `_out` variants of functions are primary functions and implemented properly as non-allocating in-place routines. And normal functions without `out=` keyword, should be implemented using the `_out` variant. For `kron`, non-allocating `tensordot_out` should be implemented first.

mruberry replied:
That's true that, ideally, the out variant is the thing (as it is for many other ops). I appreciate your concern about performance but I don't think we need to be performance sensitive on this path that leads to a runtime error.

mruberry commented on torch/_tensor_docs.py:
```diff
@@ -1892,6 +1892,13 @@ def add_docstr_all(method, docstr):
 
+add_docstr_all('kron',
+               r"""
+kron(tensor2) -> Tensor
```
"tensor2" -> "other"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3561,6 +3561,38 @@ def merge_dicts(*dicts):
+
+Computes the Kronecker product of :attr:`input` and :attr:`other`.
+
+If :attr:`input` is a :math:`(n \times m)` tensor, :attr:`other` is a
```
Very small nit: "tensor, other is a" -> "tensor and other is a"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3561,6 +3561,38 @@ def merge_dicts(*dicts):
+Supports real and complex inputs.
+
+Arguments:
+    input (Tensor): the first tensor to be multiplied
```
They're not really "multiplied." This can just leave things as input (Tensor) and other (Tensor) without elaboration.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3561,6 +3561,38 @@ def merge_dicts(*dicts):
+Keyword args:
+    out (Tensor, optional): The output tensor. Ignored if ``None``. Default: ``None``
+
+Returns:
```
Kill this "Returns:" bit. We typically only use it when the returned value is more complicated, like a tuple of multiple tensors.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3561,6 +3561,38 @@ def merge_dicts(*dicts):
+Computes the Kronecker product of :attr:`input` and :attr:`other`.
+
+If :attr:`input` is a :math:`(n \times m)` tensor, :attr:`other` is a
+:math:`(k \times l)` tensor, the result will be a :math:`(n*k \times m*l)` tensor.
```
This documentation is great. What about adding one more sentence and a mathematical statement explaining what, exactly, the Kronecker product is? I thought the first picture in its Wikipedia article (see https://en.wikipedia.org/wiki/Kronecker_product) was small and descriptive.

IvanYashchuk replied:
Sure, I've added the definition as in Wikipedia. I hope it renders correctly

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3561,6 +3561,38 @@ def merge_dicts(*dicts):
+Example::
+
+    >>> mat1 = torch.eye(2)
+    >>> mat2 = torch.ones(2, 2)
```
Reshaping a tensor created from arange might be more interesting, since it'd show distinct values in the second tensor (1, 2, 3, 4).

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2132,6 +2132,11 @@
 
+- func: kron(Tensor self, Tensor other) -> Tensor
+  variants: function, method
+
```
Per our recent dispatch updates, I think this should get an explicit "Math" dispatch key because it's implemented as a composite of existing PyTorch operators.

IvanYashchuk replied:
If no dispatch key is added then it is assumed implicitly that the function is of type "Math".

> To use autograd support for training, simply skip adding a dispatch section or write
```
dispatch:
  Math: kernel
```
> You're done. This will allow this op to be correctly registered for both inference and training.

So do we add the dispatch explicitly here?

mruberry replied:
Let's go ahead and add it. The new dispatch system is so new I'm not sure if the standard will be to elide Math dispatches in the future or not, but I'd prefer we give people examples of how it's meant to be used for now.

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2132,6 +2132,11 @@
+- func: kron(Tensor self, Tensor other) -> Tensor
+  variants: function, method
+
+- func: kron.out(Tensor self, Tensor other, *, Tensor(a!) out) -> Tensor(a!)
```
Same dispatch note here.

mruberry reviewed (commented):
This is excellent work, @IvanYashchuk, and I think it's very close to being done. I made a few comments/suggestions that should be easy tweaks.

IvanYashchuk committed:
tensor2 -> other

IvanYashchuk committed:
Updated documentation

Added mathematical definition of the Kronecker product.
Added another example.

IvanYashchuk committed:
Removed returns block in docs

IvanYashchuk committed:
Added dispatch: Math

IvanYashchuk committed:
Use at::native::resize_output in kron_out

IvanYashchuk committed:
Added non-contiguous test cases

ethanhs commented:
Hi, this is really exciting to see, I was hoping to use the kronecker product with complex tensors, but I couldn't discern if that would be supported by this. I look forward to using this!

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-kron

IvanYashchuk committed:
Use bmatrix instead of array in matrix example

IvanYashchuk committed:
Fixed non-contiguous tests

IvanYashchuk commented:
Updated `_out` implementation to use at::native::resize_output.
Explicit `dispatch: Math` is now used `native_functions.yml`.
Added non-contiguous test cases.
Documentation now includes the mathematical definition of the Kronecker product as on the Wikipedia page (tested locally that it renders correctly now).

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1689,5 +1689,71 @@ Tensor chain_matmul(TensorList matrices) {
+*/
+Tensor kron(const Tensor& self, const Tensor& other) {
+  /*
+  We can obtain the kron result using tensordot or einsum.
```
The below just uses tensordot, so maybe this should drop the "or einsum"?

IvanYashchuk replied:
It may help understanding for people not familiar with tensordot but familiar with einsum and Einstein notation (I think einsum is more common). And the next sentence uses einsum.

mruberry replied:
How about something like "using tensordot or einsum. The approach below uses tensordot."?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3594,6 +3594,51 @@ def merge_dicts(*dicts):
+
+Computes the Kronecker product, denoted by :math:`\otimes`, of :attr:`input` and :attr:`other`.
+
+If :attr:`input` is a :math:`(m \times n)` tensor and :attr:`other` is a
```
We didn't discuss this previously, but is the Kronecker product defined if either of A or B aren't matrices?

IvanYashchuk replied:
I think the Kronecker product is defined mathematically only for matrices. We can think of vectors as m×1 matrix, and scalars as 1×1 matrix, then everything works.
Vectors are tested [here](https://github.com/pytorch/pytorch/pull/45358/files#diff-7e17421f32124016eb8de04dc2f445da5786a28355e1addc72b305466f590180R202-R204) (as `(4, )` shape).
As for n-dimensional arrays with n>2, NumPy extends the definition [as described in the notes section](https://numpy.org/doc/stable/reference/generated/numpy.kron.html) to "blocks of the second tensor scaled by the first tensor".
So kron does not support batching.

Sometimes it's said that for matrices Kronecker product == tensor outer product, but this is not true for tensors in general. For [the example from Wiki about the tensor product](https://en.wikipedia.org/wiki/Outer_product#The_outer_product_of_tensors), kron would give a tensor with dimensions (3*1, 5*10, 7*100).

mruberry replied:
Being like NumPy seems OK (our goal is to be compatible, after all). Would you add a note about how input and other are treated if they're not matrices? The current docs deal with them as if they must be matrices. Something like:

- NOTE
- The Kronecker product is typically defined only for two matrices
- When either is a scalar or vector it's unsqueezed as...
- When either is an input is a tensor with 3+ dimensions then...

ALTERNATIVELY you could expand the description of the function describe the matrix case, say that scalars and vectors are unsqueezed to be matrices, and THEN define the "general" case. That seems like a more challenging but better approach.

What are your thoughts, @IvanYashchuk?

IvanYashchuk replied:
What about in the main description present the general definition first and then mention what it does for matrices.

Computes the Kronecker product of `input` and `other`.
For general n-dimensional tensors this function computes:
... math expression
<a href="https://imgbb.com/"><img src="https://i.ibb.co/Jm3ghVS/image.png" alt="image" border="0"></a><br />

The number of dimensions of `input` and `other` is assumed to be the same and if necessary the smaller tensor is unsqueezed as the larger one.

If `input` is a `(m \times n)` ~~tensor~~ matrix and `other` is a `(p \times q)` ~~tensor~~ matrix, the result will be a `(p*m \times q*n)` block ~~tensor~~ matrix:
... kron definition for matrices from wiki
Scalar and vector inputs are unsqueezed to be matrices.

mruberry replied:
That approach sounds good. Instead of 

"The number of dimensions of input and other is assumed to be the same and if necessary the smaller tensor is unsqueezed as the larger one."

I think it can say "If one tensor has fewer dimensions than the other it is unsqueezed until it has the same number of dimensions." 

That change would make the explicit reference to scalar and vector inputs at the end redundant, so it can be removed.

The last caveat is that this needs to define the dot operator in both equations above.

IvanYashchuk replied:
Okay, dot was for the normal multiplication of scalars. Is asterisk (*) preferred?

mruberry replied:
Dot's OK if the documentation defines it (the dot operator is used for so many mathematical operations it's highly ambiguous), an asterisk would also be fine and probably doesn't require definition (it's typically used for elementwise multiplication and scalar multiplication).

IvanYashchuk replied:
I've updated the documentation so that the main text describes now the computation for general n-dimensional tensors. I left the matrix case description as a note section.
Here is the image of rendered docs:
<a href="https://ibb.co/WyMZ3Qr"><img src="https://i.ibb.co/C7Ygtqj/image.png" alt="image" border="0"></a>

IvanYashchuk committed:
Updated comment in LinearAlgebra.cpp

IvanYashchuk committed:
Updated docs for general n-dim tensors

IvanYashchuk committed:
Changed 'block tensor' -> 'tensor'

IvanYashchuk committed:
Fix too long line

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3594,6 +3594,64 @@ def merge_dicts(*dicts):
+Supports real-valued and complex-valued inputs.
+
+.. note::
+    The Kronecker product is commonly defined only for matrices.
```
Want to suggest a slight tweak here to highlight the relationship between the generalized Kronecker product this implements and the definition for two matrices. Let me know what you think. 

"The Kronecker product is commonly defined only for matrices. If ..." -> "This function generalizes the typical definition of the Kronecker product for two matrices to two tensors, as described above. When :attr:\`input\` is..."

IvanYashchuk replied:
Yes, it looks better with your suggestion. I've changed the sentence now.

mruberry reviewed (approved):
Awesome work, @IvanYashchuk. Made one last small docs comment for your review.

It'll be great to have torch.kron available. I was just talking to a PyTorch user the other day about how he's using matrices with a "Kronecker structure" as a form of structured sparsity.

Do me a favor, though, and when the updates are ready for review be sure to re-request review or say "This is ready for another review." With the number of PRs I'm tracking it's hard to understand when updates are being made vs. something is ready for review again.  

Just ping me when you'd like this merged.

IvanYashchuk committed:
Changed the note according to Mike's suggestion

IvanYashchuk commented:
Hi @mruberry, I think now this PR should be ready for merging.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-kron

IvanYashchuk commented:
Tests seem to fail. Is it related to `common_methods_invocations.py`'s `method_tests`?

mruberry commented:
> Tests seem to fail. Is it related to `common_methods_invocations.py`'s `method_tests`?

I would ignore the "Facebook internal" build signal. It's complete nonsense.
